File: 20_有效的括号.cpp
#include <iostream>
#include <string.h> 
#include <vector>
#include <stack>
using namespace std;

class Solution {
    stack<char> stack;
public:
    bool isMatched(char pre, char back) {
        if (pre == '(' && back == ')')
        {
            return true;
        }
        if (pre == '[' && back == ']')
        {
            return true;
        }
        if (pre == '{' && back == '}')
        {
            return true;
        }
        return false;
    }
    bool isValid(string s) {
        char c;
        char top;
        for (int i = 0; i < s.length(); i++)
        {
            c = s[i];
            if (c == '[' || c == '{' || c == '(')
            {
                cout << "push: " << c << endl;
                stack.push(c);
            }
            else {
                if (stack.empty())
                {
                    return false;
                }
                
                top = stack.top();
                if (isMatched(top, c))
                {
                    cout << "pop: " << top << endl;
                    stack.pop();
                }
                else 
                    return false;
            }
        }
        if (stack.empty())
        {
            return true;
        }
        return false;
    }
};

int main() {
    string input;
    cin >> input;
    Solution s;
    cout << s.isValid(input) << endl;
    return 0;
}

File: 20_有效的括号.py
class Stack:
    def __init__(self):
        self.items = []
    
    def isEmpty(self):
        return self.items == []
    
    def push(self, item):
        self.items.append(item)
    
    def pop(self):
        return self.items.pop()
    
    def peek(self):
        return self.items[-1]
    
    def size(self):
        return len(self.items)

class Solution:
    def isValid(self, s: str) -> bool:
        stack = Stack()
        cnt_yuan = 0
        cnt_fang = 0
        cnt_da = 0

        length = len(s)
        for i in s:
            if i == '(':
                stack.push('(')
            if i == '[':
                stack.push('[')
            if i == '{':
                stack.push('{')
            if i == ')':
                if stack.isEmpty():
                    return False
                c = stack.pop()
                if c != '(':
                    return False
            if i == ']':
                if stack.isEmpty():
                    return False
                c = stack.pop()
                if c != '[':
                    return False
            if i == '}':
                if stack.isEmpty():
                    return False
                c = stack.pop()
                if c != '{':
                    return False
        if stack.isEmpty():
            return True
        else:
            return False

        



if __name__ == '__main__':
    s = input()
    print(Solution().isValid(s))


'''
class Solution:
    def isValid(self, s: str) -> bool:
        if not s:
            return True
        stack = []
        for i in s:
            if i in ['(', '{', '[']:
                stack.append(i)
            else:
                if not stack:
                    return False
                if i == ')' and stack[-1] == '(':
                    stack.pop()
                    continue
                if i == '}' and stack[-1] == '{':
                    stack.pop()
                    continue
                if i == ']' and stack[-1] == '[':
                    stack.pop()
                    continue
                else:
                    return False
        if stack:
            return False
        return True
'''